onnxscript/onnx_opset_builder.py

```diff
@@ -0,0 +1,375 @@
+        yield ParamDef("self")
+
+        for input in schema.inputs:
+            yield ParamDef(
+                input.name,
+                self._make_union_typeref(input.types),
+                doc=input.description)
+
+        for attr in sorted(
+                schema.attributes.values(),
+                key=lambda a: not a.required):
+            param_type = parse_attr_type(attr.type)
+            default_value = None
+
+            if attr.required:
+                pass
+            elif attr.default_value.name:
+                default_value = get_attribute_value(attr.default_value)
+
+                def fmt(value: Any) -> str:
+                    if isinstance(value, (bytes, bytearray)):
+                        return str(value.decode("utf-8"))
+                    return value
+
+                if isinstance(default_value, list):
+                    default_value = [fmt(val) for val in default_value]
+                else:
+                    default_value = fmt(default_value)
+            else:
+                default_value = NoneTypeRef
+
+            yield ParamDef(
+                attr.name,
+                param_type,
+                default_value,
+                attr.description)
+
+    def _make_union_typeref(self, onnx_types: list[str]) -> UnionTypeRef:
+        return TypeRef.make_composite_if_many(
+            UnionTypeRef,
+            *[parse_input_output_type(type) for type in sorted(onnx_types)])
+
+    def write(self, base_path: Path) -> list[Path]:
+        return sorted([self._write_module(base_path, module)
+                       for module in self.all_modules])
+
+    def _write_module(self, base_path: Path, module: ModuleDef) -> Path:
+        from os import makedirs
+        if isinstance(module, OpsetModuleDef):
+            base_path = base_path.joinpath("_impl")
+        makedirs(base_path, exist_ok=True)
+        path = base_path.joinpath(module.name + ".py")
+        with open(path, "w", encoding="utf-8") as writer:
+            self._write_header(writer)
+            module.write(writer)
+        return path
+
+    def _write_header(self, writer: TextIO):
+        from sys import argv
+        writer.write(f"# {'-' * 74}\n")
+        writer.write("# ⚠️ WARNING - AUTO-GENERATED CODE - DO NOT EDIT ⚠️ \n")
+        writer.write(f"# ⚙️ Generated by: {str.join(' ', argv)}\n")
+        writer.write(f"# {'-' * 74}\n")
+        writer.write("# Copyright (c) Microsoft Corporation. ")
+        writer.write("All rights reserved.\n")
+        writer.write("# Licensed under the MIT License.\n")
+        writer.write(f"# {'-' * 74}\n\n")
+
+
+def parse_input_output_type(onnx_type: str) -> TypeRef:
+    try:
+        id = ""
+        stack: list[TypeRef] = []
+        for c in onnx_type:
+            if c == "(":
+                if id == "tensor":
+                    type = TensorTypeRef()
+                elif id == "seq":
+                    type = SequenceTypeRef()
```

Comment:
One question we need to resolve is whether we should use distinct type (annotation) for attributes and values. My instinct is to differentiate them, but maybe we might be able to get by without doing it for Sequences. It may also tie in with what warnings IDEs give, relating to ONNX Scripts restrictions (though, we may want to relax them eventually: eg., supporting the use of `[X, Y, Z]` for a sequence of tensors.)